src/rust/lib/graph-editor/src/component/edge.rs

```diff
@@ -44,23 +47,280 @@ const NODE_PADDING       : f32 = node::SHADOW_SIZE;
+    fn snap_to_shape_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>>;
+    fn sprite(&self) -> &Sprite;
+
+    /// Return the angle perpendicular to the shape at the given point. Defaults to zero, if not
+    /// implemented.
+    fn cut_angle_for_point(&self, point:Vector2<f32>) -> nalgebra::Rotation2<f32>  {
+        let local = self.to_shape_coordinate_system(point);
+        self.cut_angle_for_point_local(local)
+    }
+
+    fn cut_angle_for_point_local(&self, _point:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+        nalgebra::Rotation2::new(0.0)
+    }
+
+    /// Snaps the coordinates given in the global coordinate system to the shape and returns
+    /// the global coordinates. If no snapping is possible, returns `None`.
+    fn snap_to_shape(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+        let local          = self.to_shape_coordinate_system(point);
+        let local_snapped  = self.snap_to_shape_local(local)?;
+        let global_snapped = self.from_shape_to_global_coordinate_system(local_snapped);
+        Some(global_snapped)
+    }
+
+    /// Set the hover split for this shape. The `split_data` indicates where the shape should be
+    /// split and which side should be recolored.
+    fn enable_hover_split(&self, split_data:SplitData) {
+        let SplitData {split_position,area,cut_angle} = split_data;
+        // Compute rotation in shape local coordinate system.
+        let base_rotation        = self.display_object().rotation().z;
+        let hover_split_rotation = cut_angle + base_rotation;
+        match area {
+            Area::Above => self.set_hover_split_rotation(hover_split_rotation),
+            Area::Below => {
+                self.set_hover_split_rotation(hover_split_rotation + 2.0 * RIGHT_ANGLE)
+            },
+        }
+        // Compute position in shape local coordinate system.
+        let offset = self.to_shape_coordinate_system(split_position);
+        self.set_hover_split_offset(offset)
+    }
+
+    /// Disable the hover split on this shape.
+    fn disable_hover_split(&self) {
+        self.set_hover_split_offset(Vector2::new(INFINITE, INFINITE));
+        self.set_hover_split_rotation(RIGHT_ANGLE);
+    }
+
+    /// Convert the given global coordinates to the shape local coordinate system.
+    fn to_shape_coordinate_system(&self, point:Vector2<f32>) -> Vector2<f32> {
+        let base_rotation   = self.display_object().rotation().z;
+        let local_unrotated = point - self.display_object().global_position().xy();
+        UnitComplex::new(-base_rotation) * local_unrotated
+    }
+
+    /// Convert the given shape local coordinate to the global coordinates system.
+    fn from_shape_to_global_coordinate_system(&self, point:Vector2<f32>) -> Vector2<f32> {
+        let base_rotation   = self.display_object().rotation().z;
+        let local_unrotated = UnitComplex::new(base_rotation) * point;
+        local_unrotated + self.display_object().global_position().xy()
```

Comment:
Please, do not use explicit `.display_object()` ever. All functions are exported there, so `self.global_position()` should work. Fix it everywhere and in case you will find use case where you need something more, lets talk. Explicit call like here should always be avoided.

Reply:
Implementing `ObjectOps` and avoiding the call to `.display_object()` seems to mess with my ability to return a reference to the `EdgeShape`. For now, I would prefer to leave it as is, instead of tracking down why `ObjectOps` requires a static lifetime only to avoid this method call.